server/svix-server/src/core/types.rs

```diff
@@ -526,36 +526,72 @@ pub struct ExpiringSigningKey {
     pub expiration: DateTime<Utc>,
 }
 
+const FORBIDDEN_KEYS: [&str; 19] = [
+    "user-agent",
+    "keep-alive",
+    "proxy-authenticate",
+    "proxy-authorization",
+    "te",
+    "trailers",
+    "transfer-encoding",
+    "upgrade",
+    "age",
+    "cache-control",
+    "clear-site-data",
+    "expires",
+    "pragma",
+    "warning",
+    "content-length",
+    "content-type",
+    "content-encoding",
+    "content-language",
+    "content-location",
+];
+
+const FORBIDDEN_PREFIXES: [&str; 10] = [
+    "x-amz-", "x-amzn-", "x-google", "x-goog-", "x-gfe", "x-amz-", "x-azure-", "x-fd-", "x-svix-",
+    "svix-",
+];
+
+fn validate_header_key(k: &str, errors: &mut ValidationErrors) {
+    let k = &k.to_lowercase();
+    if let Err(_e) = http::header::HeaderName::try_from(k) {
+        errors.add(ALL_ERROR, ValidationError::new("Invalid Header Name."));
+    }
+    if FORBIDDEN_KEYS.contains(&k.as_str()) {
+        errors.add(
+            ALL_ERROR,
+            ValidationError::new("Header uses a forbidden key."),
+        );
+    }
+    FORBIDDEN_PREFIXES.iter().for_each(|p| {
+        if k.starts_with(p) {
+            errors.add(
+                ALL_ERROR,
+                ValidationError::new("Header starts with a forbidden prefix."),
+            )
+        }
+    })
+}
+
 #[derive(Clone, Debug, PartialEq, Serialize, Default)]
 pub struct EndpointHeaders(pub HashMap<String, String>);
 json_wrapper!(EndpointHeaders);
 
 impl EndpointHeaders {
-    const FORBIDDEN_KEYS: &'static [&'static str] = &[
-        "user-agent",
-        "keep-alive",
-        "proxy-authenticate",
-        "proxy-authorization",
-        "te",
-        "trailers",
-        "transfer-encoding",
-        "upgrade",
-        "age",
-        "cache-control",
-        "clear-site-data",
-        "expires",
-        "pragma",
-        "warning",
-        "content-length",
-        "content-type",
-        "content-encoding",
-        "content-language",
-        "content-location",
-    ];
-    const FORBIDDEN_PREFIXES: &'static [&'static str] = &[
-        "x-amz-", "x-amzn-", "x-google", "x-goog-", "x-gfe", "x-amz-", "x-azure-", "x-fd-",
-        "x-svix-", "svix-",
-    ];
+    pub fn new() -> Self {
+        EndpointHeaders(HashMap::new())
+    }
+}
+
+impl FromIterator<(String, String)> for EndpointHeaders {
```

Comment:
Why is this needed all of a sudden?

Reply:
This is so that I can construct a new `EndpointHeaders` type that is compatible with the SeaORM model, from the `EndpointHeadersPatch` type.

Reply:
But why wasn't it needed before your changes?

Reply:
`EndpointHeadersPatch` did not exist before this PR, and it is also incompatible with the SeaORM model. This implementation exists so I can map `EndpointHeadersPatch` to `EndpointHeaders` for SeaORM.

Reply:
This is not for Patch, this is for EndpointHeaders which did exist before. That's my confusion. Why is it only needed for something that existed already and not for the new addition.

Reply:
It is required [here](https://github.com/svix/svix-webhooks/blob/8eac6be878bc8641f97ab8176e1b38bc70a6c691/server/svix-server/src/v1/endpoints/endpoint/mod.rs#L241).

Reply:
Why wasn't it needed before though? That's the confusion.

Reply:
Following up on our convo: we can't collect into `EndpointHeadersIn` unless we do the same implementation for _that_ type.